### PR TITLE
Feature/tests: 100% method coverage

### PR DESCRIPTION
**Pull request by hugocowan**
Created tests that provide **100% coverage** of all bar methods. This is to make sure that small errors that cause errors in the methods can be caught before they are deployed. If the methods pass the tests, it means they are working.

Added test packages to the `build.gradle` file. You may need to restart your IDE once the changes have been pulled down to get these packages installed.

Added a dedicated `EasyBlastFurnacePluginRunner` file to separate actual tests from the executable method. You may need to edit your build configuration file to point to this new file.

Moved all string and int variables to do with methods to the enums `/utils/Strings` and `/utils/CoalPer`. This means they only exist in one place and so can be changed more easily, and it also means these variables are now accessible for testing.

Removed sections from `MetalBarMethod` and `GoldHybridMethod` that the tests showed to be unused.

Tested every single method thoroughly & manually to ensure the tests are accurate.

image

### CODE REVIEW

File: src/main/java/com/toofifty/easyblastfurnace/utils/Strings.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    WITHDRAW_STAMINA_POTION4("Withdraw stamina potion"),
+    DRINK_STAMINA_POTION1("Drink stamina potion"),
+    DRINK_STAMINA_POTION2("Drink stamina potion"),
+    DRINK_STAMINA_POTION3("Drink stamina potion"),
+    DRINK_STAMINA_POTION4("Drink stamina potion"),
+    GET_MORE_STAMINA_POTIONS("Get more stamina potions! Check settings to disable this");
+
+
+    @Getter
+    private final String txt;
```
Comment (Toofifty) [inline src/main/java/com/toofifty/easyblastfurnace/utils/Strings.java:62]: Scared of the letter e? 😆 

I'm not much of a java guy but I don't think there's much benefit to using an enum here - a class with static consts will make usage a bit neater IMO

Comment (hugocowan) [reply src/main/java/com/toofifty/easyblastfurnace/utils/Strings.java:62]: Haha, yeah I'm not much of one myself but I'm doing my best! I'll make it a class with static consts

File: src/main/java/com/toofifty/easyblastfurnace/methods/GoldHybridMethod.java [hunk 2]
```diff
@@ -49,11 +49,6 @@ public MethodStep next(BlastFurnaceState state)
         // continue doing gold bars until enough coal has been deposited
         // then do one trip of metal bars
 
-        if (state.getInventory().has(ItemID.GOLD_ORE) &&
-            !state.getEquipment().equipped(ItemID.GOLDSMITH_GAUNTLETS, ItemID.SMITHING_CAPE, ItemID.SMITHING_CAPET)) {
-            return equipGoldsmithGauntlets;
-        }
```
Comment (Toofifty) [inline src/main/java/com/toofifty/easyblastfurnace/methods/GoldHybridMethod.java:55]: Correct me if I'm wrong, but I think this one is needed to make sure the player has the gold gauntlets equipped before they drop them in the hopper (if they have ice gloves currently equipped)

Comment (hugocowan) [reply src/main/java/com/toofifty/easyblastfurnace/methods/GoldHybridMethod.java:55]: I'll double check, I tested it but maybe I only tested with a smithing cape and not with the goldsmith gauntlets

File: src/main/java/com/toofifty/easyblastfurnace/utils/Strings.java (hunk 1 above)
Comment (hugocowan) [reply src/main/java/com/toofifty/easyblastfurnace/utils/Strings.java:62]: Okay, so a reason I made it an enum was so I could use the names of the enum variables as well as the values, with lines such as `Strings.MITHRIL.name()`

I'm working out how to do something similar with a class and static consts, but it looks like it's going to be a little messy.

For example, this:

`@Test
    public void adamantiteBarMethod()
    {
        metalMethods(ItemID.ADAMANTITE_ORE, ItemID.ADAMANTITE_BAR, BarsOres.ADAMANTITE_BAR.name(), Strings.ADAMANTITE.name());
        assertEquals(Strings.WITHDRAWADAMANTITEORE.getTxt(), methodHandler.getStep().getTooltip());
    }`

becomes this if I make Strings a class with static consts:
`    @Test
    public void adamantiteBarMethod() throws NoSuchFieldException
    {
        metalMethods(ItemID.ADAMANTITE_ORE, ItemID.ADAMANTITE_BAR, BarsOres.ADAMANTITE_BAR.name(), Strings.class.getField("ADAMANTITE").getName());
        assertEquals(Strings.WITHDRAWADAMANTITEORE, methodHandler.getStep().getTooltip());
    }`

File: src/test/java/com/toofifty/easyblastfurnace/EasyBlastFurnacePluginTest.java [hunk 3]
```diff
@@ -1,13 +1,364 @@
 package com.toofifty.easyblastfurnace;
```
Comment (Toofifty) [inline src/test/java/com/toofifty/easyblastfurnace/EasyBlastFurnacePluginTest.java:1]: Still can't believe you added tests! I'm more than happy to pull these in as-is :) 

My suggestion though would be to split them up a bit more and make them a bit _dumber_ - having if statements in tests usually means your tests are doing too much. Same with having assertions extracted outside of the test method body IMO. Though extraction of _setup_ tasks is definitely helpful for readability.

Here's how I think I'd structure one of these:

```java
public void mithrilHybridMethod()
{
    setInventoryCount(ItemID.MITHRIL_ORE, 1);
    setInventoryCount(ItemID.GOLD_ORE, 1);
    assertEquals(Strings.MITHRIL_BAR_METHOD, methodHandler.getMethod().getName());

    assertStepTooltip(Strings.WITHDRAW_COAL_BAG);

    setInventoryCount(ItemID.COAL_BAG, 1);
    // or even something like withdrawCoalBag();
    assertStepTooltip(Strings.FILL_COAL_BAG);

    fillCoalBag();
    assertStepTooltip(Strings.WITHDRAW_ICE_OR_SMITHS_GLOVES);

    setInventoryCount(ItemID.ICE_GLOVES, 1);
    setBankCount(ItemID.SMITHING_CAPE, 1);
    assertStepTooltip(Strings.WITHDRAW_SMITHING_CAPE);

    // ...
}

Comment (hugocowan) [reply src/test/java/com/toofifty/easyblastfurnace/EasyBlastFurnacePluginTest.java:1]: Okay, I'll do my best to make it look more like that

File: src/main/java/com/toofifty/easyblastfurnace/utils/Strings.java (hunk 1 above)
Comment (hugocowan) [reply src/main/java/com/toofifty/easyblastfurnace/utils/Strings.java:62]:
```Java
assertEquals(Strings.valueOf(barMethod).getTxt(), methodHandler.getMethod().getName());
```

There's also this line where I get the value of a variable in Strings using the dynamic argument `barMethod`. This seems very difficult to do with a class.

File: src/main/java/com/toofifty/easyblastfurnace/methods/GoldHybridMethod.java (hunk 2 above)
Comment (hugocowan) [reply src/main/java/com/toofifty/easyblastfurnace/methods/GoldHybridMethod.java:55]: Yep you do need that, that's my bad 😬 I really need to stop testing with the smithing cape

File: src/test/java/com/toofifty/easyblastfurnace/EasyBlastFurnacePluginTest.java (hunk 3 above)
Comment (hugocowan) [reply src/test/java/com/toofifty/easyblastfurnace/EasyBlastFurnacePluginTest.java:1]: @Toofifty I've done my best to do as you suggested. Please take a look and let me know if you want me to do anything else.

File: src/main/java/com/toofifty/easyblastfurnace/utils/Strings.java (hunk 1 above)
Comment (hugocowan) [reply src/main/java/com/toofifty/easyblastfurnace/utils/Strings.java]: I rewrote the tests in the way you suggested, and that allowed me to make this a class without any issues.

**Comment by hugocowan** [comment]
I added a new change to the PR along with the tests. I've made it so the player is instructed to "Deposit Inventory" if they're carrying metal (not gold or coal) ore if there's not enough coal in the furnace for them. The key bits of code for this are in `MetalBarMethod` and `GoldHybridMethod`, here are the relevant snippets:

```Java
if (state.getInventory().has(ItemID.COAL, ItemID.GOLD_ORE) **|| (state.getInventory().has(oreItem()) && !coalRun)**) {
            return putOntoConveyorBelt;
}
```

```Java
if (state.getInventory().has(ItemID.GOLD_BAR, barItem()**, oreItem()**)) {
                return depositInventory;
}
```

I bolded the bits that are new. I think this is a worthwhile change, it stops noobs from putting, say, runite ore in when there's no coal and having to collect a few bars at a time for the next few coal runs. The tests have been updated to accommodate this and are still at 100% coverage for methods.

**Comment by hugocowan** [comment]
@Toofifty Please let me know if there's anything I can do to help get this merged. I know it's a lot to look through.

**Review by Toofifty — APPROVED** [review]
LGTM @hugocowan 🙌

**Pull request merged**